### ISSUE TITLE
Make tests parallelizable

### ISSUE THREAD

**Issue by pawamoy** [qtile/qtile#1634]
I looked at the test suite code, and it *looks* like tests should be able to run in parallel, though as soon as I use 2 workers with `pytest-xdist`, tests start failing randomly.

In `conftest.Xephyr.start_xephyr()`:

```python
self.display = ":{}".format(xcffib.testing.find_display())
```

`xcffib.find_display()`:

```python
def find_display():
    display = 10
    while True:
        if not os.path.exists(lock_path(display)):
            return display
        display += 1
```

Doesn't look very thread-safe to me? I mean, is `os.path.exists` atomic?

Also, not sure if 10 is a hard-limit or an arbitrary value.

**Comment by tych0** [comment]
Yes, it is. I just copied this code from lightdm, note the comment: https://github.com/canonical/lightdm/blob/master/src/dm-tool.c#L389

I suppose we could change xcffib to use flock, so at least it wouldn't race with itself?

**Comment by pawamoy** [comment]
Something like

```python
import fnctl

def find_display():
    display = 10
    while True:
        lockfile = lockpath(display)
        if not os.path.exists(lockfile):
            os.system("touch %s" % lockfile)
            f = open(lockfile, "r")
            try:
                fcntl.flock(f.fileno(), fcntl.LOCK_EX|fcntl.LOCK_NB)
                # successfully got a lock
                return display
            except Exception:
                # failed to get a lock
                pass  # try next one
        display += 1
```

Bits found at https://www.programcreek.com/python/example/1605/fcntl.flock

**Comment by tych0** [comment]
Yeah, exactly. I'd definitely take a patch for xcffib that looked like that.

**Comment by pawamoy** [comment]
I can't get it to work.

```python
def lock_path(display):
    return '/tmp/.X%d-lock' % display


def get_lock(display):
    lockfile = lock_path(display)
    if os.path.exists(lockfile):
        print("file exists")
        return False
    try:
        os.system("touch %s" % lockfile)
        f = open(lockfile, "r")
        fcntl.flock(f.fileno(), fcntl.LOCK_EX | fcntl.LOCK_NB)
        return True  # successfully got a lock
    except Exception:
        return False  # failed to get a lock


def release_lock(display):
    lockfile = lock_path(display)
    if not os.path.exists(lockfile):
        print("file does not exist")
        return True
    try:
        f = open(lockfile, "r")
        # the unlock operation doesn't raise any exception
        # so instead we try to re-acquire the lock, then release it
        fcntl.flock(f.fileno(), fcntl.LOCK_EX | fcntl.LOCK_NB)
        fcntl.flock(f.fileno(), fcntl.LOCK_UN)
    except Exception:
        return False  # failed to release the lock
    else:
        f.close()
        os.remove(lockfile)
        return True
```

I can release a lock obtained in process 1, from within process 2.

Not sure to understand the behavior. Info at https://stackoverflow.com/questions/28470246/python-lockf-and-flock-behaviour

I'm a bit confused by the fact we are using both a file *and* a lock on this file.

Since we create a file anyway, couldn't we just use `os.mkdir()`?

```python
def lock_path(display):
    return '/tmp/.X%d-lock' % display


def get_lock(display):
    lockfile = lock_path(display)
    try:
        os.mkdir(lockfile)
        return True
    except FileExistsError:
        return False
```

Then in `stop_xephyr` we could simply `os.rmdir(lock_path(self.display))`.

**Comment by pawamoy** [comment]
Or really we don't bother with releasing flock locks since it seems they are released when the process terminates. We will keep increasing the display value, and

- two tests from the same worker could use the same value, but sequentially, so it should not be an issue
- two tests from different workers could not use the same value since there's a lock

The created files should still be removed at the end of the test of course.

You tell me :smile:!

- flock with no releasing (only file removal)
- or mkdir and rmdir

**Comment by tych0** [comment]
I'm confused; why do you need to re-acquire a lock you already have when you're trying to release it?

It is expected behavior that task 2 can release a lock that task 1 acquired; everyone has to be speaking the same protocol. I just don't understand why you'd ever do that (intentionally or otherwise) in this case.

**Comment by pawamoy** [comment]
Because I wanted to handle the case where the lock couldn't be released. But I guess it never happens then? So it's indeed useless!

**Comment by pawamoy** [comment]
- The `XvfbTest` class from `xcffib` is setting `os.environ["DISPLAY"]` to an unused display.
- In Qtile's `conftest.py`, the `qtile` fixture uses the `xephyr` fixture which uses the `xvfb` fixture.
- The `xvfb` fixture uses `XvfbTest` in a `with` clause.
- But then the `Xephyr` class tries to get a *new* display with

```python
self.display = ":{}".format(xcffib.testing.find_display())
```

I think it should simply grab it from `os.environ` since the `xvfb` fixture must have set it correctly.

Unless I don't understand how pytest fixtures work:

```python
@pytest.fixture(scope="session")
def xvfb():
    with xcffib.testing.XvfbTest():
        display = os.environ["DISPLAY"]
        if not can_connect_x11(display):
            raise OSError("Xvfb did not come up")

        yield


@pytest.fixture(scope="function")
def xephyr(request, xvfb):
    kwargs = getattr(request, "param", {})

    with Xephyr(**kwargs) as x:
        yield x

# the contents of the xephyr fixture runs inside the with clause of the xvfb one, right?
```

The Xephyr manager is also trying to release the lock itself:

```python
        # clean up the lock file for the display we allocated
        try:
            os.remove(xcffib.testing.lock_path(int(self.display[1:])))
        except OSError:
            pass
```

But it shouldn't, since the xvfb manager is already releasing it!

Anyway, I tried with that change, and print the display, but no luck, still getting these captured outputs from pytest:

```
------------------------------------------------ Captured stdout setup ------------------------------------------------
:10
------------------------------------------------ Captured stderr setup ------------------------------------------------
(EE) 
Fatal server error:
(EE) Server is already active for display 10
	If this server is no longer running, remove /tmp/.X10-lock
	and start again.
(EE)
```

**Comment by pawamoy** [comment]
I'll play with all this in https://github.com/pawamoy/flock-playground and report back when I have something working. I might ask questions on IRC as well!

**Comment by ramnes** [comment]
I'm randomly hitting that same `xcffib.ConnectionException` when running the test suite, without even using pytest-xdist.

**Comment by tych0** [comment]
Have you tried 0.10.1? That should have the right locking and such.

**Comment by ramnes** [comment]
Yes, I'm already using xcffib 0.10.1. :(

**Comment by vrusinov** [comment]
I can also reproduce this without parallelisation - see https://github.com/qtile/qtile/issues/2103 and while using xcffib 0.11.1.

**Comment by yobleck** [comment]
Sorry for necro bumping, but I was messing around with pytest-parallel
as an alternative to pytest-xdist and I figured I'd post some initial results here:

hack to fix working with python 3.9 (fixed in next release?):
https://github.com/browsertron/pytest-parallel/pull/98/files

cpu=ryzen 3700x
baseline ~10-12min? 2runs no fails
workers=2 ~6min 2runs no fails
workers=4 ~3min 7runs no fails 1 frozen test? test_main
workers=6 ~2min 4runs some fails at test_memory_units/defaults
workers=8 ~1min 1run lots of fails mostly to do with asyncio and GenPollUrl widgets

More testing needed but 2-4 workers might be a good balance.

**Comment by tych0** [comment]
The more failures at higher thread counts suggest that we have tests that are racy, and we're just not hitting those races at lower thread levels. Racy tests means unreliable CI, which is very important to avoid. Seems like we'd need to fix all such races before we parallelize the tests.

**Comment by elParaguayo** [comment]
Interesting. So the list of failures from the higher number of workers may give us some pointers.
